envoy/api/v2/auth/rbac.proto

```diff
@@ -0,0 +1,180 @@
+
+  // Optional. A list of HTTP methods (e.g., "GET", "POST").
+  // It is ignored in gRPC case because the value is always "POST".
+  // If set to ["*"] or not specified, it applies to any method.
+  repeated string methods = 3;
+
+  // Definition of a custom constraint. The fields used in ServiceRole definition map to attributes
+  // defined in AttributeContext
+  // (https://github.com/envoyproxy/data-plane-api/blob/master/envoy/service/auth/v2/attribute_context.proto).
+  // The key of a constraint refers to the attribute name in AttributeContext. For example,
+  // destination.labels["app"], request.http.headers["version"].
+  message Constraint {
+    // Key of the constraint.
+    string key = 1;
+
+    // List of valid values for the constraint.
+    // Exact match, prefix match, and suffix match are supported for constraint values.
+    // For example, the value "v1alpha2" matches
+    // "v1alpha2" (exact match), or "v1*" (prefix match),
+    // or "*alpha2" (suffix match).
+    repeated string values = 2;
+  }
+
+  // Optional. Extra constraints in the ServiceRole specification.
+  // The above ServiceRole examples shows an example of constraint "version".
+  repeated Constraint constraints = 4;
+}
+
+// ServiceRoleBinding assigns a ServiceRole to a list of subjects.
+// This represents the "Spec" part of the ServiceRoleBinding object. The name and namespace
+// of the ServiceRoleBinding is specified in "metadata" section of the ServiceRoleBinding
+// object.
+message ServiceRoleBinding {
+  // Required. List of subjects that are assigned the ServiceRole object.
+  repeated Subject subjects = 1;
+
+  // Required. Reference to the ServiceRole object.
+  RoleRef role_ref = 2 [(validate.rules).message.required = true];
+}
+
+// Subject defines an identity or a group of identities. The identity is either a user or
+// a group or identified by a set of "properties". The names of a property
+// used to identify a subject in ServiceRoleBinding map to attributes defined in AttributeContext
+// (https://github.com/envoyproxy/data-plane-api/blob/master/envoy/service/auth/v2/attribute_context.proto)
+// as follows:
+//    user: source.principal
+//    group is currently undefined. It is reserved to be used in the future.
+//
+// Custom property name directly refers to the attribute name in AttributeContext. For example,
+// source.service, source.labels["app"], request.http.headers["version"].
+message Subject {
+  // Optional. The user name/ID that the subject represents.
+  string user = 1;
+
+  // Optional. The group that the subject belongs to.
+  string group = 2;
+
+  // Optional. The set of properties that identify the subject.
+  map<string, string> properties = 3;
+}
+
+// RoleRef refers to a role object.
+message RoleRef {
+  // Required. The type of the role being referenced.
+  // Currently, "ServiceRole" is the only supported value for "kind".
+  string kind = 1 [(validate.rules).string.const = "ServiceRole"];
```

Comment:
can we use an enum here?

Reply:
Based on API style naming convention (https://cloud.google.com/apis/design/naming_convention#enum_names), "Enum values must use CAPITALIZED_NAMES_WITH_UNDERSCORES". But I want to use "ServiceRole" constant here, because it needs to match ServiceRole definition "kind: ServiceRole".

Reply:
The name is arbitrary (when serialized enums are represented as integers). We aren't doing string comparison between these are we?

Reply:
You don't even need this, you can always add this later or upgrade to `oneof` without breaking backward compatibility. I'd quote [STYLE.md](https://github.com/envoyproxy/envoy/blob/master/STYLE.md) again that prefers `oneof` than enum + string.

Reply:
ok, let me remove "kind" for now. I can add it later if needed.